### ISSUE TITLE
Source changes after the first aren't pushed to the cluster when using skaffold dev

### ISSUE THREAD

**Issue by ekuefler** [GoogleContainerTools/skaffold#5855]
### Expected behavior

Whenever I make a change to my source code while running `skaffold dev`, Skaffold should re-build my images and then update my deployments to use the new image.

### Actual behavior

The first change I make works as intended, replacing the image from my manifests with the image built by skaffold. But later updates are ignored - Skaffold builds the new image and claims to deploy it, but I can see that the image specified in the deployment hasn't changed.

### Information

- Skaffold version: v1.23.0
- Operating system: Linux
- Installed via: skaffold.dev
- Contents of skaffold.yaml:

```yaml
apiVersion: skaffold/v2beta14
kind: Config
metadata:
  name: environments
build:
  tagPolicy:
    sha256: {}
  artifacts:
  - image: gcr.io/devops-environments/manager
    docker:
      dockerfile: cmd/manager/Dockerfile
deploy:
  kubectl:
    remoteManifests:
    - environments-system:deployment/manager
```

### Steps to reproduce the behavior

1. Run `skaffold dev` using the above `skaffold yaml`
2. Make a source change
3. Observe that skaffold performs a build and updates the image in the deployment
4. Make another source change
5. Observer that skaffold performs a build, but the deployment is not updated.

### Notes

I suspect this has something to do with `sha256` or `remoteManifests` in that Skaffold isn't able to identify the image that needs to be updated after it's changed it once. I see this in the log:

```
Tags used in deployment:
 - gcr.io/devops-environments/manager -> gcr.io/devops-environments-playground/gcr.io/devops-environments/manager:latest@sha256:e3c25ea096bd7c01d649267605c34ecddc3df32d4e51370507d56ae7fb5f83c2
```

But after the first change, `gcr.io/devops-environments/manager` is no longer in the yaml, so perhaps it's not able to do the replacement.

**Comment by ekuefler** [comment]
Also worth noting: if I manually edit the deployment to go back to the original image, it's then able to detect the next change again.

**Comment by tejal29** [comment]
Interesting, is it possible to run skaffold with `-v=debug` and provide debug logs?
Feel free to attach a paste.

**Comment by gsquared94** [comment]
I was able to repro this issue. It happens with non-local cluster like GKE where skaffold applies the kubernetes manifest with images referenced by digest. In an ongoing devloop, using `kubectl.remoteManifests` when skaffold re-reads the updated manifest from the cluster it finds that images are referenced by digest and skips over them in the next deployment.
https://github.com/GoogleContainerTools/skaffold/blob/7a2ba7760078f30fa959e7eb81487ebdabb74876/pkg/skaffold/kubernetes/manifest/images.go#L107-L110

I'm not sure why skaffold originally decided to not overwrite images specified by digest in a kubernetes manifest (see https://github.com/GoogleContainerTools/skaffold/pull/2773 about issue https://github.com/GoogleContainerTools/skaffold/issues/834). However this outright breaks `skaffold dev` for `kubectl.remoteManifests`.

**Comment by briandealwis** [comment]
The underlying problem is: how does `dev` clean up after the change?

One way to address this would be to download a local copy of the remote manifest and apply changes to that manifest. On cleanup, we redeploy this original manifest.

**Comment by gsquared94** [comment]
> The underlying problem is: how does `dev` clean up after the change?

do you mean in-between dev-loops or at the end of the `skaffold dev` process?

> One way to address this would be to download a local copy of the remote manifest and apply changes to that manifest. On cleanup, we redeploy this original manifest.

I thought this is what it is doing today.
https://github.com/GoogleContainerTools/skaffold/blob/f95999ffad973fce7f3a9fbd2d96354fc2eb0749/pkg/skaffold/deploy/kubectl/kubectl.go#L306-L311

**Comment by briandealwis** [comment]
> do you mean in-between dev-loops or at the end of the skaffold dev process?

At the end of the dev process.

From this bug, it seems we have issues with both during and after the process.

**Comment by srntqn** [comment]
@briandealwis hi! Do you have any plans to fix this behaviour? 

Got this problem using skaffold v1.27.0 and GKE.

**Comment by srntqn** [comment]
I guess, the problem here is not only with images referenced by digest:
 https://github.com/GoogleContainerTools/skaffold/blob/7a2ba7760078f30fa959e7eb81487ebdabb74876/pkg/skaffold/kubernetes/manifest/images.go#L107-L110
 
 I've removed it from source code, rebuild skaffold binary and was able to update images using `skaffold run` and GKE. 
 
 But if I use `--default-repo` option the problem appears again.  If I get it right @ekuefler uses this option too.
 
 Looks like `skaffold` doesn't handle image name correctly after the first run with `--default-repo` option.  
 
 For example, I have this configuration:
 
 ```yaml
 apiVersion: skaffold/v2beta11
kind: Config
build:
  tagPolicy:
    dateTime:
      format: "2006-01-02_15-04-05.999_MST"
      timezone: "Local"
  artifacts:
  - image: eu.gcr.io/test-project/skaffold-example
deploy:
  kubectl:
    remoteManifests:
    - pod/getting-started
 ```
 
 `getting-started` pod in cluster uses this image: `eu.gcr.io/test-project/skaffold-example:v1.27.0-9-g823896bdd-dirty`
 
 Then I run following `skaffold` command:
 
```bash
skaffold run -v trace --default-repo europe-west1-docker.pkg.dev/test-images/local-dev-images
```

Updated pod now uses this image:

```bash
europe-west1-docker.pkg.dev/test-images/local-dev-images/eu.gcr.io/test-project/skaffold-example
```
It's okay according to the [docs](https://skaffold.dev/docs/environment/image-registries/).

I change the code and  then run `skaffold` to update the pod in the cluster:

```bash
skaffold run -v trace --default-repo europe-west1-docker.pkg.dev/test-images/local-dev-images
```

In logs I can see this line:

```bash
DEBU[0011] image [eu.gcr.io/test-project/skaffold-example] is not used by the current deployment
```

`skaffold` doesn't know about previous run when we changed image name using `--default-repo` so it continues to use image name from the `skaffold.yaml`  and does nothing with manifest in the cluster because now it uses another image.
